heka/sandbox/decoders/http_edge_decoder.lua

```diff
@@ -62,20 +76,43 @@ local msg = {
 }
 
 function process_message()
+    -- First, copy the current message as-is.
+    landfill_msg.Timestamp  = read_message("Timestamp")
+    landfill_msg.Type       = read_message("Type")
+    landfill_msg.Hostname   = read_message("Hostname")
+    landfill_msg.Pid        = read_message("Timestamp")
+    -- UUID is auto-generated and meaningless anyways
+    landfill_msg.Logger     = read_message("Logger")
+    landfill_msg.Payload    = read_message("Payload")
+    landfill_msg.EnvVersion = read_message("EnvVersion")
+    landfill_msg.Severity   = read_message("Severity")
+    -- Now copy the fields:
+    landfill_msg.Fields = {}
+    while true do
+        local value_type, name, value, representation, count = read_next_field()
+        if not name then break end
+        -- Keep the first occurence only (we want the value supplied by the
+        -- HttpListenInput, not the user-supplied one if we have to choose).
+        if not landfill_msg.Fields[name] then
+            landfill_msg.Fields[name] = value
+        end
+    end
+    inject_message(landfill_msg)
+
     -- Reset Fields, since different namespaces may use different fields.
-    msg.Fields = {}
+    main_msg.Fields = {}
```

Comment:
Then I would at least move the data from the landfill message to the new message instead of hitting the CGO interface again.